bgzf/bgzf.go

```diff
@@ -0,0 +1,108 @@
+package bgzf
+
+import (
+	"bufio"
+	"bytes"
+	"compress/gzip"
+	"log"
+	"math"
+	"os"
+)
+
+// Reader is the decompressor for BGZF files which is at it's core
+// a gzip reader. The BGZF reader retains a *os.File to use for
+// seek functionality as well as a *bufio.Reader to implement the
+// io.ByteReader interface, required for reading gzip files as
+// independent concatenated blocks, rather than a multistream.
+type Reader struct {
+	file         *os.File
+	buf          *bufio.Reader
+	decompressor *gzip.Reader
+}
+
+// Block represents a single BGZF block stores as a bytes.Buffer
+// initialized to have a capacity of 65536 (max uint16).
+type Block struct {
+	bytes.Buffer // cap == 2^16
+}
+
+// NewBlock returns a pointer to a bytes buffer with capacity 65536.
+func NewBlock() *Block {
+	b := new(Block)
+	b.Grow(math.MaxUint16)
+	return b
+}
+
+// Offset for a location in the BGZF file consisting of the compressed
+// offset (the number of bytes before the beginning of the desired block
+// starting from the beginning of the file), and and the uncompressed
+// offset (the number of bytes from the desired record starting from the
+// beginning of the uncompressed block)
+type Offset struct {
+	Compressed   int64  // offset in compressed file
+	Uncompressed uint16 // offset in uncompressed block
+}
+
+// NewReader opens a new BGZF reader for the input file
+func NewReader(filename string) Reader {
+	var r Reader
+	var err error
+	r.file, err = os.Open(filename)
+	if err != nil {
+		log.Panic(err)
+	}
+
+	r.buf = bufio.NewReader(r.file) // for ReadByte method
+
+	r.decompressor, err = gzip.NewReader(r.buf)
+	if err != nil {
+		log.Panic(err)
+	}
+
+	r.decompressor.Multistream(false)
```

Comment:
Is this needed here?  I see it also at the start of ReadBlock.  I am wondering if because it is always called at the beginning of ReadBlock, maybe it doesn't need to be called here.

Reply:
You are correct. I will remove that. The multistream was a pain to figure out and I think I got a bit trigger-happy with it.